### PR TITLE
Fix for crash in core validation handling of destroyed command pools

### PR DESCRIPTION
**Pull request by jzulauf-lunarg**
Address crash reported #720 - Possible crash in layers when deleting command pools https://vulkan.lunarg.com/issue/view/59eeea8166315153977b0fe6

Fix more robustly cleans up validation meta-data when deleting command buffer pools.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -4088,6 +4088,24 @@ static bool checkCommandBuffersInFlight(layer_data *dev_data, COMMAND_POOL_NODE
     return skip;
 }
 
+// Free all command buffers in given list, removing all references/links to them using resetCB
+static void FreeCommandBufferNodes(layer_data *dev_data, COMMAND_POOL_NODE *pPool, const uint32_t commandBufferCount,
```
Comment (tobine) [inline layers/core_validation.cpp]: Kill camelCase on var names.

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -4164,49 +4169,40 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateQueryPool(VkDevice device, const VkQueryPoo
...
-            if (fb_state) fb_state->cb_bindings.erase(cb_node);
-        }
-        dev_data->commandBufferMap.erase(cb);  // Remove this command buffer
-        delete cb_node;                        // delete CB info structure
+static void PostCallRecordDestroyCommandPool(layer_data *dev_data, VkCommandPool pool) {
+    COMMAND_POOL_NODE *cp_state = GetCommandPoolNode(dev_data, pool);
+    // Remove cmdpool from cmdpoolmap, after freeing layer data for the command buffers
+    // "When a pool is destroyed, all command buffers allocated from the pool are freed."
+    if (cp_state) {
+        // Create a vector, as freeCommandBufferList deletes from cp_node->commandBuffers during iteration.
```
Comment (tobine) [inline layers/core_validation.cpp]: Some wrong names in this comment?
freeCommandBufferList -> FreeCommandBufferNodes?
cp_node -> cp_state?

File: tests/layer_validation_tests.cpp [hunk 3]
```diff
@@ -20325,6 +20325,59 @@ TEST_F(VkPositiveLayerTest, DeleteDescriptorSetLayoutsBeforeDescriptorSets) {
...
+    VkCommandBufferBeginInfo secondary_begin = {};
+    secondary_begin.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+    secondary_begin.pNext = NULL;
+    secondary_begin.flags = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
+    secondary_begin.pInheritanceInfo = &cmd_buf_inheritance_info;
+
+    res = vkBeginCommandBuffer(secondary_cmds, &secondary_begin);
+    ASSERT_VK_SUCCESS(res);
+    vkEndCommandBuffer(secondary_cmds);
+    ASSERT_VK_SUCCESS(res);
```
Comment (tobine) [inline tests/layer_validation_tests.cpp]: Unneeded assert

File: tests/layer_validation_tests.cpp [hunk 4]
```diff
@@ -20325,6 +20325,59 @@ TEST_F(VkPositiveLayerTest, DeleteDescriptorSetLayoutsBeforeDescriptorSets) {
...
+
+    m_commandBuffer->begin();
+    vkCmdExecuteCommands(m_commandBuffer->handle(), 1, &secondary_cmds);
+    m_commandBuffer->end();
+
+    // DestroyCommandPool *implicitly* frees the command buffers allocated from it
+    vkDestroyCommandPool(m_device->handle(), secondary_cmd_pool, NULL);
+    // If bookkeeping has be lax, validating the reset will attempt to touch deleted data
```
Comment (tobine) [inline tests/layer_validation_tests.cpp]: "has be" = "has been"?

File: layers/core_validation.cpp [hunk 5]
```diff
@@ -1793,7 +1793,7 @@ static void removeCommandBufferBinding(layer_data *dev_data, VK_OBJECT const *ob
 }
 // Reset the command buffer state
 //  Maintain the createInfo and set state to CB_NEW, but clear all other state
-static void resetCB(layer_data *dev_data, const VkCommandBuffer cb) {
+static void ResetCommandBufferNode(layer_data *dev_data, const VkCommandBuffer cb) {
```
Comment (tobine) [inline layers/core_validation.cpp]: Note that the "Node" name in general was determined to be non-informative and we're trying to transition "node" references to "state" references.

Comment (mark-lunarg) [reply layers/core_validation.cpp]: Note that if you do any node-state or other renames across the file, to put them in a separate commit.

**Pull request merged**